### ISSUE TITLE
onPositionDiscontinuity is not getting called when AdGroup is reached.

### ISSUE THREAD

**Issue by sravan1213** [google/ExoPlayer#8456]
onPositionDiscontinuity is not getting called when AdGroup. This is happening for content where subtitles are side loaded.

1. Side load subtitle for a media with midroll ads
2. Wait for the midroll position to hit
3. Observe player state remains unchanged. Ad UI wont appear and even trackers are not fired. The entire ad behavior get messed up.

used following assets to test.
content - _https://bitdash-a.akamaihd.net/content/sintel/hls/playlist.m3u8_
subtitle url to sideload - _https://bitdash-a.akamaihd.net/content/sintel/hls/subtitles_en.vtt_ (same subtitle from content)

[video_bugreport_demoPatch.zip](https://github.com/google/ExoPlayer/files/5807531/video_bugreport_demoPatch.zip)

**Comment by sravan1213** [comment]
@andrewlewis can you please check this or provide some pointers to where to look at ? 

Thanks in advance.

**Comment by vinodkputta** [comment]
@ojw28  @andrewlewis We are facing this issue on the production app. Can someone please help us to fix this issue? 

Thanks.

**Comment by andrewlewis** [comment]
Sorry for the delay getting to this. I added this sample to the demo app samples list:

```
{
  "name": "Issue #8456",
  "uri": "https://bitdash-a.akamaihd.net/content/sintel/hls/playlist.m3u8",
  "ad_tag_uri": "https://pubads.g.doubleclick.net/gampad/ads?sz=640x480&iu=/124319096/external/ad_rule_samples&ciu_szs=300x250&ad_rule=1&impl=s&gdfp_req=1&env=vp&output=vmap&unviewed_position_start=1&cust_params=deployment%3Ddevsite%26sample_ar%3Dpremidpostoptimizedpodbumper&cmsid=496&vid=short_onecue&correlator=",
  "subtitle_uri": "https://bitdash-a.akamaihd.net/content/sintel/hls/subtitles_en.vtt",
  "subtitle_mime_type": "text/vtt",
  "subtitle_language": "en"
},
```

... replicating the "MIDROLL test" item from the patch in the linked archive, but it played correctly in the demo app using the latest `dev-v2` version. Please could you try doing the same and checking whether the ad UI shows for you? Note: I didn't apply the cookie-related parts of the patch but I'm not sure how those could be related.

**Comment by sravan1213** [comment]
Thanks for the reply @andrewlewis 

Have you checked the video I attached to repro the issue ?
We have to select side loaded subtitles from text tracks. I was able to reproduce 100% on 2.12.2 & 2.12.3

Cookie is not related to this issue :)

I'll check on dev branch and let you know. Thanks.

**Comment by sravan1213** [comment]
@andrewlewis I am able to reproduce on dev-v2 branch as well. Repro rate is 100%

Let me know if you need more details. 
Make sure you select the side loaded subtitle is selected before mid roll hits.
And let the ad play as user is watching. Do not seek.

**Comment by sravan1213** [comment]
@andrewlewis were you able to reproduce this ??

**Comment by sravan1213** [comment]
@andrewlewis I've done some debugging, when side-loaded subtitle is enabled, 
`renderer.isEnded()` is always return false for text track causing this issue. I'm not able to figure out why is this happening.

**Comment by hyl0376** [comment]
I also reproduced this issue in 2.13.1 demo app.
The weird thing is that 'https://storage.googleapis.com/exoplayer-test-media-1/webvtt/numeric-lines.vtt' works fine but 'https://bitdash-a.akamaihd.net/content/sintel/hls/subtitles_en.vtt' doesn't.
sample setting :
```
      {
        "name": "subtitle pre-mid-post",
        "uri": "https://content.jwplatform.com/manifests/vM7nH0Kl.m3u8",    // not matter
        "ad_tag_uri": "https://pubads.g.doubleclick.net/gampad/ads?sz=640x480&iu=/124319096/external/ad_rule_samples&ciu_szs=300x250&ad_rule=1&impl=s&gdfp_req=1&env=vp&output=vmap&unviewed_position_start=1&cust_params=deployment%3Ddevsite%26sample_ar%3Dpremidpost&cmsid=496&vid=short_onecue&correlator=",    // VMAP pre-, mid- and post-rolls, single ads
        "subtitle_uri": "https://bitdash-a.akamaihd.net/content/sintel/hls/subtitles_en.vtt",
        "subtitle_mime_type": "text/vtt",
        "subtitle_language": "en"
      }
```
reproduction step : 
1. pre-roll play
2. before mid-roll play (15s), click 'SELECT TRACKS' at the top left and select text track 'Auto' to 'english' (have to change in 'SELECT TRACKS', not in controller)
3. wait for the mid-roll play (not seeking to or over mid-roll)
4. mid-roll play with no overlay and controller is still enabled (no onPositionDiscontinuity() event)

I faced the same problem and it's very critical.
Please take a look.

**Comment by icbaker** [comment]
I can reproduce what you're seeing using the instructions in https://github.com/google/ExoPlayer/issues/8456#issuecomment-783149153.

I believe this happens because the subtitles aren't correctly 'ended' when transitioning into the ad period. I'm not sure exactly why that is yet, I'm going to investigate further. It's likely related to the fact that sideloaded subtitles generally don't work correctly with ClippingMediaPeriod (which is used by midroll ad insertion): https://github.com/google/ExoPlayer/issues/5317 

However even with that limitation the current behaviour is unexpected.

> The weird thing is that 'https://storage.googleapis.com/exoplayer-test-media-1/webvtt/numeric-lines.vtt' works fine but 'https://bitdash-a.akamaihd.net/content/sintel/hls/subtitles_en.vtt' doesn't.

I think this is because numeric-lines.vtt is short, so playback of the subtitle content is fully completed before the midroll appears.

I tried another experiment with subtitle content both before, during and after the midroll and I found the subtitles just keep playing throughout the ad. This means the subtitles after the midroll are completely out of sync with the content (by the duration of the midroll). It's also unrelated to the content being streamed over HLS - it reproduces the same with progressive mp4 content:

```json
      {
        "name": "subtitles before and after mid-roll (progressive)",
        "uri": "https://storage.googleapis.com/exoplayer-test-media-1/gen-3/screens/dash-vod-single-segment/video-avc-baseline-480.mp4",
        "ad_tag_uri": "https://pubads.g.doubleclick.net/gampad/ads?sz=640x480&iu=/124319096/external/ad_rule_samples&ciu_szs=300x250&ad_rule=1&impl=s&gdfp_req=1&env=vp&output=vmap&unviewed_position_start=1&cust_params=deployment%3Ddevsite%26sample_ar%3Dpremidpost&cmsid=496&vid=short_onecue&correlator=",
        "subtitle_uri": "https://storage.googleapis.com/exoplayer-test-media-1/ssa/test-subs-position.ass",
        "subtitle_mime_type": "text/x-ssa",
        "subtitle_language": "en"
      }
```

**Comment by icbaker** [comment]
I can also reproduce the same problem without using ads, just by concatenating a clipped MergingMediaSource with subtitles and a ProgressiveMediaSource:

```java
DefaultDataSourceFactory dataSourceFactory = new DefaultDataSourceFactory(this);
String videoUri =
    "https://storage.googleapis.com/exoplayer-test-media-1/gen-3/screens/dash-vod-single-segment/video-avc-baseline-480.mp4";
String subtitlesUri =
    "https://storage.googleapis.com/exoplayer-test-media-1/ssa/test-subs-position.ass";
MergingMediaSource mergingMediaSource =
    new MergingMediaSource(
        new ProgressiveMediaSource.Factory(dataSourceFactory)
            .createMediaSource(MediaItem.fromUri(videoUri)),
        new SingleSampleMediaSource.Factory(dataSourceFactory)
            .createMediaSource(
                new MediaItem.Subtitle(
                    Uri.parse(subtitlesUri),
                    MimeTypes.TEXT_SSA,
                    "en",
                    C.SELECTION_FLAG_DEFAULT),
                C.TIME_UNSET));
ConcatenatingMediaSource concatenatingMediaSource =
    new ConcatenatingMediaSource(
        new ClippingMediaSource(
            mergingMediaSource, /* startPositionUs= */ 0, /* endPositionUs= */ 5_000_000),
        new ProgressiveMediaSource.Factory(dataSourceFactory)
            .createMediaSource(MediaItem.fromUri("https://html5demos.com/assets/dizzy.mp4")),
        new ClippingMediaSource(
            mergingMediaSource,
            /* startPositionUs= */ 5_000_000,
            /* endPositionUs= */ C.TIME_END_OF_SOURCE));
player.setMediaSource(concatenatingMediaSource);
```

At 5 seconds the second video starts playing, but the subtitles continue to be rendered on top of it. This is unexpected. At the end of the second video, it seems to pause and the subtitles continue to render until the subtitles are complete, at which point the first video plays again (starting from 5s in) with subtitles.

**Comment by nishantkp** [comment]
We are also facing this issue with the production app. Do you have an ETA for this one?

**Comment by sravan1213** [comment]
@icbaker do we have any eta on this.

Apologies for bugging. Thanks in advance.

**Comment by icbaker** [comment]
I have a fix for this in review - when it's submitted it will appear linked to this issue.

**Comment by nishantkp** [comment]
hi @icbaker, do you have an ETA for the upcoming release?

**Comment by icbaker** [comment]
This fix will be included in the 2.14.0 release, which is tentatively scheduled for the beginning of May. It is likely to also be included in future 2.13.X minor releases, if we do any.

**Comment by hyl0376** [comment]
Hi @icbaker, could you release it earlier?
This issue is critical to our product release schedule and May is too late.

**Comment by nishantkp** [comment]
I agree with @hyl0376, if it is possible to do a minor release, it would be great since is May is too late.

**Comment by icbaker** [comment]
If you need this change more urgently than the next ExoPlayer release you can depend on a local copy of the library: https://github.com/google/ExoPlayer/blob/dev-v2/README.md#locally

Then you can either use the dev-v2 branch at tip-of-tree, or create your own local branch that cherrypicks the commit above on top of 2.13.2.

**Comment by icbaker** [comment]
This fix is included in the 2.13.3 release that was just published.

**Comment by nishantkp** [comment]
Thank you so much @icbaker
